Fix text selection toolbar alignment for RTL languages

SalehTZ opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

**Fix text selection toolbar alignment for RTL (right-to-left) languages**

This PR updates the Material text selection toolbar to properly support right-to-left (RTL) layouts. Previously, the toolbar and its buttons did not align or order correctly when using RTL text direction, leading to a broken or confusing user experience for RTL languages.

**Summary of changes:**
- Refactored layout logic in `text_selection_toolbar.dart` to handle RTL and LTR directions for both horizontal and overflow (vertical) arrangements.
- Updated padding logic in `text_selection_toolbar_text_button.dart` to use `EdgeInsetsDirectional`, ensuring correct padding for both text directions.
- Added new widget tests to verify that:
  - Toolbar items are ordered right-to-left in RTL contexts.
  - The overflow menu behaves correctly in RTL, showing items in the correct order and position.

**Before:**  
- Toolbar items and overflow menus did not respect RTL layout, resulting in misaligned or reversed button order.
- Padding was inconsistent in RTL.
<details>
<summary>Before Screenshots</summary>

| LTR Not Expanded | LTR Expanded | RTL Not Expanded | RTL Expanded |
|---|---|---|---|
| ![before_ltr_non_expanded jpg](https://github.com/user-attachments/assets/bccf719e-c95a-46b1-b924-d2fb229e4ada) | ![before_ltr_expanded jpg](https://github.com/user-attachments/assets/a947690a-62fb-4f41-863c-d3822294bb66) | ![before_rtl_non_expanded](https://github.com/user-attachments/assets/1e3440ee-d80c-4554-aeda-7632f2b166ca) | ![before_rtl_expanded](https://github.com/user-attachments/assets/364a3b2b-d068-4488-afbd-61224e74c30a) |

</details>


**After:**  
- Toolbar items and overflow menus are correctly ordered and aligned for RTL.
- Button padding is now directionally correct.

<details>
<summary>After Screenshots</summary>

| LTR Not Expanded | LTR Expanded | RTL Not Expanded | RTL Expanded |
|---|---|---|---|
| ![after_ltr_non_expanded](https://github.com/user-attachments/assets/b63bf625-0d37-4304-9ba5-af017c197b15) | ![after_ltr_expanded](https://github.com/user-attachments/assets/6a95c9b7-e343-4e7e-93ae-f18de6d38a2e) | ![after_rtl_not_expanded](https://github.com/user-attachments/assets/8a93b08b-fcf2-45a9-8f5a-390133c0aec4) | ![after_rtl_expanded](https://github.com/user-attachments/assets/5cbdffd4-3943-44cc-98d9-fa470e11b942) |

</details>

This improves the usability and accessibility of text selection controls for users of RTL languages.


Fixes #59306 

---

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

SalehTZ committed:
refactor: change padding type to EdgeInsetsGeometry for better flexibility for both rtl and ltr

SalehTZ committed:
fix: RTL TextDirection in TextSelectionToolbarItemsLayout

This change ensures that the Material text selection toolbar correctly positions its items (including the overflow button) according to the current `TextDirection` (LTR or RTL).

Previously, the layout logic did not account for text direction, leading to incorrect placement of items in RTL layouts.

The `_TextSelectionToolbarItemsLayout` and its corresponding `_RenderTextSelectionToolbarItemsLayout` now explicitly take a `textDirection` parameter. The `_placeChildren` method in the render object has been updated to arrange items from right-to-left when in an RTL context.

SalehTZ committed:
feat: added tests tp ensure TextSelectionToolbar items are ordered right-to-left in RTL

This commit introduces tests to verify that:
- Items in `TextSelectionToolbar` are correctly ordered from right-to-left when the `textDirection` is `TextDirection.rtl`.
- The overflow menu in `TextSelectionToolbar` functions as expected in RTL mode, correctly displaying and hiding overflowed items.

SalehTZ committed:
feat: added tests to ensure TextSelectionToolbar items are ordered right-to-left in RTL

This commit introduces tests to verify that:
- Items in `TextSelectionToolbar` are correctly ordered from right-to-left when the `textDirection` is `TextDirection.rtl`.
- The overflow menu in `TextSelectionToolbar` functions as expected in RTL mode, correctly displaying and hiding overflowed items.

SalehTZ committed:
Merge remote-tracking branch 'origin/fix-text-selection-toolbar-in-rtl' into fix-text-selection-toolbar-in-rtl

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar_text_button.dart:
```diff
@@ -81,21 +81,24 @@ class TextSelectionToolbarTextButton extends StatelessWidget {
   /// Standard Material [TextSelectionToolbar]s have buttons with different
   /// padding depending on their position in the toolbar.
-  static EdgeInsets getPadding(int index, int total) {
+  static EdgeInsetsGeometry getPadding(int index, int total) {
```
I've started the Google tests on this PR just to make sure this doesn't break anyone.

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -513,57 +536,145 @@ class _RenderTextSelectionToolbarItemsLayout extends RenderBox
+      double totalHorizontalWidth = 0;
+      double maxHorizontalHeight = 0;
+
+      // First pass: calculate dimensions and collect items
```
Nit: Add a period at the end of these comments here and below.

SalehTZ replied:
Done.

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -513,57 +536,145 @@ class _RenderTextSelectionToolbarItemsLayout extends RenderBox
+          itemParentData.offset = Offset(rightEdge, 0.0);
+        }
+      } else {
+        if (isRtl) {
```
It looks to me like this boolean can't be true here? I think the `else` above is the case where isRtl is false, and this whole `if` block is duplicated and can be removed.

SalehTZ replied:
Wow, You're absolutely right, fix on the way.

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -513,57 +536,145 @@ class _RenderTextSelectionToolbarItemsLayout extends RenderBox
-      // Handle placing the navigation button after iterating all children.
-      if (renderObjectChild == navButton) {
-        return;
+    if (!overflowOpen) {
```
Would it be useful to split the layout when overflowOpen true/false into two separate methods in order to reduce some of the nested conditionals in this method? I find it somewhat hard to read (though it is harder to read on GitHub than in the IDE).

justinmc commented on packages/flutter/test/material/text_selection_toolbar_test.dart:
```diff
@@ -352,4 +353,101 @@ void main() {
+      ),
+    );
+
+    // Verify all items are visible
```
Periods at the ends of these comments as well.

SalehTZ replied:
Done.

justinmc commented on packages/flutter/test/material/text_selection_toolbar_test.dart:
```diff
@@ -352,4 +353,101 @@ void main() {
+    // In RTL, items should be ordered right-to-left
+    for (int i = 0; i < itemRects.length - 1; i++) {
+      expect(
+        itemRects[i].left,
+        equals(itemRects[i + 1].right),
```
Is the order of itemRects guaranteed to be the order that they appear in the UI?

You could also `find.text('$i')` and make sure that it's right next to `find.text('${i + 1}')`.

SalehTZ replied:
for `find.text('$i')` we should pass a child to `TestBox` widget as well. That's doable but IDK if that's necessary.

With small change in `TestBox`, I can tweak the test a little bit (double check rightEdges) to be like below, is this more reliable?

<details>
<summary>Details</summary>

```dart
testWidgets('items are ordered right-to-left in RTL', (WidgetTester tester) async {
    const int itemCount = 3;
    final List<Widget> children = List<Widget>.generate(
      itemCount,
      (int i) => TestBox(key: ValueKey<String>('item_$i'), child: Text('$i')),
    );

    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: Directionality(
            textDirection: TextDirection.rtl,
            child: TextSelectionToolbar(
              anchorAbove: const Offset(50.0, 100.0),
              anchorBelow: const Offset(50.0, 200.0),
              children: children,
            ),
          ),
        ),
      ),
    );

    // Verify all items are visible.
    expect(find.byType(TestBox), findsNWidgets(itemCount));

    // Find all text widgets by their content and get their positions.
    final List<Rect> textRects = List<Rect>.generate(
      itemCount,
      (int i) => tester.getRect(find.text('$i')),
    );

    // In RTL, items should be in reverse order (2, 1, 0).
    // So item 2 should be rightmost, then 1, then 0.
    for (int i = 0; i < itemCount - 1; i++) {
      final Rect current = textRects[i];
      final Rect next = textRects[i + 1];

      // In RTL, each item should be to the left of the previous one.
      expect(
        next.right,
        lessThanOrEqualTo(current.left),
        reason: 'In RTL, item ${i + 1} should be to the left of item $i',
      );
    }

    // Verify the visual order by checking the rightmost position.
    final List<double> rightEdges = textRects.map((Rect r) => r.right).toList();
    final List<double> sortedRightEdges = List<double>.from(rightEdges)
      ..sort((double a, double b) => b.compareTo(a));
    expect(
      rightEdges,
      equals(sortedRightEdges),
      reason: 'Items should be ordered right-to-left in RTL',
    );
  });
  
```
</details>

justinmc reviewed (commented):
Thank you so much for stepping in to fix this. This is a high quality PR fixing what must be a very annoying for RTL users. I appreciate the screenshots too.

The comments below are mostly small stuff, overall this looks good.

SalehTZ committed:
fix: added missing dots to the end of docs

SalehTZ committed:
fix: added missing dots to the end of docs

SalehTZ committed:
fix: remove duplicate if clause

SalehTZ commented:
@justinmc  Is it good to add this line to `hitTestChildren` method?

```dart
// Don't hit test children aren't shown.
if (!childParentData.shouldPaint) {
    return false;
}
```

result be like:

```dart
  // Include the parent data offset in the hit test.
  @override
  bool hitTestChildren(BoxHitTestResult result, {required Offset position}) {
    // The x, y parameters have the top left of the node's box as the origin.
    final ToolbarItemsParentData childParentData = child!.parentData! as ToolbarItemsParentData;

    // Don't hit test children aren't shown.
    if (!childParentData.shouldPaint) {
      return false;
    }

    return result.addWithPaintOffset(
      offset: childParentData.offset,
      position: position,
      hitTest: (BoxHitTestResult result, Offset transformed) {
        assert(transformed == position - childParentData.offset);
        return child!.hitTest(result, position: transformed);
      },
    );
  }
```

SalehTZ committed:
Refactor: Separate TextSelectionToolbar layout logic into horizontal and vertical methods

This change refactors the `_placeChildren` method in `TextSelectionToolbar` by extracting the logic for horizontal and vertical layouts into two new methods: `_placeChildrenHorizontally` and `_placeChildrenVertically`.

This improves code readability and maintainability by clearly separating the distinct layout calculations.

SalehTZ committed:
fix: TextSelectionToolbar RTL item order test

The previous test for right-to-left (RTL) item ordering in `TextSelectionToolbar` incorrectly assumed items would be perfectly adjacent. This commit revises the test to accurately verify that items are positioned to the left of their preceding items in RTL, allowing for potential spacing or visual separation between them.

The test now:
- Adds `Text` widgets as children to `TestBox` for easier identification and position retrieval.
- Verifies that `next.right` is less than or equal to `current.left` for each pair of items in RTL.
- Additionally confirms the visual right-to-left order by comparing the sorted list of right edges with the actual list.

justinmc reviewed (commented):
Let's see what happens with CI, hopefully when you update the PR it will go green.

About your hitTestChildren question in https://github.com/flutter/flutter/pull/169854#issuecomment-2951746574. Is that the hitTestChildren in _TextSelectionToolbarTrailingEdgeAlignRenderBox? Yeah good optimization, I say add it.

SalehTZ committed:
Merge branch 'flutter:master' into fix-text-selection-toolbar-in-rtl

SalehTZ commented:
> ```dart
> // Don't hit test children aren't shown.
>     if (!childParentData.shouldPaint) {
>       return false;
>     }
> ```

Yes it is, forgot to mention the class 😁. Then i'll add it as well.

SalehTZ commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -510,63 +533,138 @@ class _RenderTextSelectionToolbarItemsLayout extends RenderBox
+    double maxHorizontalHeight = 0;
+
+    // First pass: calculate dimensions and collect items.
+    int childIdx = 0;
```
done

SalehTZ commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -510,63 +533,138 @@ class _RenderTextSelectionToolbarItemsLayout extends RenderBox
-        return;
+    if (isRtl) {
+      // In RTL, we want the nav button on the left and items right-aligned.
+      if (navButtonIsPainted) {
```
hmm, I'll check it 🤔

justinmc replied:
Yeah either way is fine with me.

SalehTZ commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -510,63 +533,138 @@ class _RenderTextSelectionToolbarItemsLayout extends RenderBox
+    // First pass: calculate dimensions and collect items.
+    int childIdx = 0;
+    RenderBox? current = firstChild;
+    while (current != null) {
```
done

SalehTZ committed:
fix: Fix typo in TextSelectionToolbar RTL test comment

The comment incorrectly stated that item 2 should be rightmost in RTL. It has been corrected to leftmost.

SalehTZ committed:
Refactor: Simplify horizontal layout logic in TextSelectionToolbar

This commit refactors the `_placeChildrenHorizontally` method in `TextSelectionToolbar` for improved clarity and efficiency.

The changes include:
- Renaming variables for better readability (e.g., `paintedContentItems` to `contentItems`, `totalHorizontalWidth` to `totalWidth`).
- Simplifying the logic for collecting and positioning child items by iterating directly over children using `visitChildren`.
- Using a boolean `showNavButton` derived from `_lastIndexThatFits` to determine if the navigation button should be painted, instead of a separate `navButtonIsPainted` flag.

SalehTZ commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -510,63 +533,138 @@ class _RenderTextSelectionToolbarItemsLayout extends RenderBox
-        return;
+      // Position content items from right to left.
+      double rightEdge = totalHorizontalWidth;
+      for (final RenderBox item in paintedContentItems) {
```
Yeah, that's a fair point. I went back and forth on that too. I ended up sticking with two loops because it felt cleaner to read. And like you said, performance isn't really an issue (specially for a small toolbar), so I figured it was a good trade-off for making it more straightforward.

for the `for` loops inside `if (isRtl) {} else {}` it would be something like this:
```dart
double currentX = 0.0;
double rightEdge = isRtl ? totalWidth : 0.0;

for (final RenderBox item in contentItems) {
  final itemParentData = item.parentData! as ToolbarItemsParentData;
  if (isRtl) {
    rightEdge -= item.size.width;
    itemParentData.offset = Offset(rightEdge, 0.0);
  } else {
    itemParentData.offset = Offset(currentX, 0.0);
    currentX += item.size.width;
  }
}
```

but overall the only potential micro-optimization would be to use a `for` loop with manual sibling traversal instead of `visitChildren` to avoid the function call overhead, but the difference would be negligible and not worth the reduced readability.

justinmc replied:
That makes sense, I agree and it's more readable so I say keep it as-is. Thanks for thoroughly investigating that.

SalehTZ committed:
Refactor: Improve layout logic in Material text selection toolbar

This commit refactors the `_placeChildrenVertically` method within the `_TextSelectionToolbarContentLayout` class of the Material text selection toolbar.

The changes include:
- Using `visitChildren` for iterating over child render objects, replacing the manual linked-list traversal.
- Renaming `maxItemWidth` to `maxWidth` for clarity.
- Removing the unused `nextSize` variable.
- Simplifying the logic for handling the navigation button.

These changes improve the readability and maintainability of the vertical layout logic for the overflow menu in the Material text selection toolbar.

SalehTZ committed:
Refactor: Simplify logic in `TextSelectionToolbar`

This commit refactors the logic within the `TextSelectionToolbar`'s `performLayout` method.
Specifically, it streamlines the condition for determining whether a child should be painted. Instead of checking `_shouldPaintChild` and then setting `shouldPaint` to true or false, it now directly checks `!_shouldPaintChild` to set `shouldPaint` to false, with an `else` block to set it to true. This improves readability without changing the underlying behavior.

SalehTZ committed:
Merge branch 'flutter:master' into fix-text-selection-toolbar-in-rtl

SalehTZ committed:
Merge branch 'flutter:master' into fix-text-selection-toolbar-in-rtl

justinmc reviewed (approved):
LGTM if Google tests pass :+1: . Thanks for jumping in to help with RTL. It's a bit of a blindspot for me unfortunately.

 > (https://github.com/flutter/flutter/pull/169854#issuecomment-2957641023) If its priority isn't high, I'll work on this later and create a separate issue and PR for it.

Sounds good. I didn't expect a change of behavior from that optimization so I must have misread it. Makes sense to do it in a separate PR.

SalehTZ commented:
That's an honor to work on Flutter, I'll contribute more, especially on RTL, currently can't work much due to missile attacks on Iran, but I'll be back soon.✌🏻

victorsanni commented on packages/flutter/test/material/text_selection_toolbar_test.dart:
```diff
@@ -352,4 +353,111 @@ void main() {
+    );
+  });
+
+  testWidgets('puts children in an overflow menu if they overflow in RTL', (
```
This test passes on `master`, but I'm assuming it is just making sure overflow menu works anyways in RTL (outside of this PR)?

victorsanni reviewed (approved):
Hi @SalehTZ, thanks for the PR, it looks really good. The detailed PR header with the screenshots make it easier to follow and review as well. Great work!

SalehTZ commented:
Thank you guys, you guys are awesome and your words meant so much to me 😍
